### PR TITLE
Nakamoto: Add stacker/signer bit vector and event

### PR DESCRIPTION
**Pull request by kantai**
### Description

This PR implements a handful of things:

1. It adds a `BitVec` class to stacks-common. This PR is at least the third implementation of a bit vector in the codebase, so it seemed like adding a simple data struct would be useful here.
2. Adds a `BitVec` field to the `NakamotoBlockHeader`, this is included in the consensus serialization and the stacker/signer signature hash. It is _not_ included in the miner's signature hash.
3. Adds a `new_pox_set` endpoint for the event dispatcher. This endpoint announces when a reward set is calculated in a PoX anchor block. It relays information about the reward set:
      * The cycle number of the set
      * The StacksBlockId of the anchor block
      * The canonically ordered list of reward addresses (with repetitions for addresses with multiple slots)
      * The canonically ordered list of signers, each entry containing a signing address (C32-encoded), the number of signing slots, and the amount stacked. This canonical ordering is what will be used for entries in the bit vector.
4. It implements `make_signer_set`, which is invoked by `make_reward_set` to translate the list of `RawRewardEntry`s into signing set entries. Currently, it does this by summing all entries with the same public key and assigning slots based on the slot thresholds.
5. It mocks data for the signing key in raw reward entries in `nakamoto_get_reward_set()`
6. It alters the logic deciding between using `nakamoto` reward set fetching and `2.x`: any reward set that is used by Nakamoto should be fetched via the `nakamoto` reward set endpoint. The first nakamoto reward set _will_ occur in epoch 2.5.
7. It fixes a broken integration test `correct_burn_outs` -- this integration test wasn't asserting that the sortition db advanced, so it missed the fact that the recent update in `next` to require signing keys (and unique ones) broke the test. The test now asserts the sortition db advances _and_ it works.
8. It update the `correct_burn_outs` test to check the results of the `new_pox_set` event observer API are correct.

### Still needs test coverage

- `make_signer_set` needs unit testing. This is easy enough to add and I will add it before this merges, but before I did that, I wanted to get feedback on this PR because if `make_signer_set` needs to change, it seems like a waste to write the unit tests now.

### Relevant issues

#4241 - this is the `stacks-node` side of tracking. this PR doesn't implement the punishment for non-participation, though.
#4256 - this PR emits this list
#4255 - summing over the signer set entries in this PR gives the amount of active lockup. the amount of total lockup (which could include people who miss the threshold) would need to be added as another field in the return. This should be simple to add and I can do it if its useful (@zone117x).

### CODE REVIEW

File: stackslib/src/chainstate/nakamoto/coordinator/mod.rs
Comment (jcnelson) [inline stackslib/src/chainstate/nakamoto/coordinator/mod.rs:71]: Is there an issue for this?  Because this seems to be pretty high-priority!

File: stackslib/src/chainstate/nakamoto/coordinator/mod.rs
Comment (jcnelson) [inline stackslib/src/chainstate/nakamoto/coordinator/mod.rs:116]: Is this truly fatal?  If so, then should it panic?

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (jcnelson) [inline stackslib/src/chainstate/nakamoto/mod.rs:320]: I'm a little skeptical that we should be using the generic `BitVec` here.  Yes, we need a bitmap of some kind, but I think we also need a hard bound on how big this is allowed to be (e.g. there can be no more than 4000 bits).  And, I think that bound needs to be expressed in the data type itself, so it's impossible to represent a block header that would exceed this bound (and thus it's impossible to both serialize and deserialize such a block).

File: stackslib/src/chainstate/stacks/boot/mod.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/mod.rs:630]: Yes

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (jcnelson) [inline testnet/stacks-node/src/tests/nakamoto_integrations.rs:1007]: Any way to expand this to e.g. 10 stackers, some of which don't make the stacking minimum?

**Review by jcnelson — COMMENTED** [review]
I think this is on the right track.  Just a few comments.

File: stackslib/src/chainstate/coordinator/mod.rs
Comment (zone117x) [reply stackslib/src/chainstate/coordinator/mod.rs:315]: Yeah it seems like this code contradicts a line from the PR description:
> The first nakamoto reward set will occur in epoch 2.5.

Comment (kantai) [reply stackslib/src/chainstate/coordinator/mod.rs:315]: > Should we be checking at epoch 2.5? Are all epoch 2.5 reward sets Nakamoto reward sets?

I don't think that they must be or will be. There's a couple reasons for this. First, there's basically two differences between nakamoto reward sets and 2.x reward sets: (1) they _must_ contain signing keys, and so the code asserts as much, and (2) if there's no participation, an error is yielded and burnchain processing halts.

For the first constraint, you can see that there will always be at least one reward cycle in 2.5 which didn't use pox-4. Imagine reward cycle 10 starts at block height 300, and epoch 2.5 starts at block height 300, then reward cycle 10 would be in epoch-2.5, but there's no way it could be a nakamoto reward set: it cannot load signing keys because it could not have been constructed in pox-4.

The second difference is even more important, though, because non-participation during 2.5 _doesn't_ mean a chain halts. It only means a chain halts if the reward-cycle is also part of epoch-3.0. This could be true of the last cycle in 2.5 (but no other ones).

> Yeah it seems like this code contradicts a line from the PR description:
>> The first nakamoto reward set will occur in epoch 2.5.

I think what's happening here is a little confusing -- let's say Epoch-3.0 starts at height 305 and each reward cycle is length 10. Reward cycle 30 is the first nakamoto reward set, even though its start height is in Epoch 2.5. What's important is that every reward set which is active at any time in Epoch 3.0 be a nakamoto reward set. So that's what this calculation is doing: figure out what cycle is active at epoch-3's start height and then find the cycle start height of that cycle.

File: stackslib/src/chainstate/nakamoto/coordinator/mod.rs
Comment (kantai) [reply stackslib/src/chainstate/nakamoto/coordinator/mod.rs:71]: There's a lot of ongoing discussion of this in the open .signers PR, I'm not sure that it has its own issue yet, but its definitely top of mind there.

File: stackslib/src/chainstate/nakamoto/coordinator/mod.rs
Comment (kantai) [reply stackslib/src/chainstate/nakamoto/coordinator/mod.rs:116]: I think its as fatal as non-participation, so it should be treated the same.

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (kantai) [reply stackslib/src/chainstate/nakamoto/mod.rs]: I added a generic const to the `BitVec` struct for `MAX_SIZE`. The field in the block header sets this to 4000.

File: stackslib/src/chainstate/stacks/boot/mod.rs
Comment (kantai) [reply stackslib/src/chainstate/stacks/boot/mod.rs:630]: I'd argue this is purely a code-aesthetic choice, really. This branch is only reached if the `signing_principal` set in the pox-4 contract is a contract principal, and the important thing is just that the address is unusable, which the burn address will be regardless of version bit.

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (kantai) [reply testnet/stacks-node/src/tests/nakamoto_integrations.rs:1007]: Yes, but I think that kind of thing is better handled by a unit test in the `make_signer_set` code. Integration tests are very cumbersome for that (not just to write, but to run and maintain) because it will definitely flap if its making very tight assertions (and it should). This is exactly the unit tests that I was waiting to add until I'd gotten some initial feedback on the approach.

File: stackslib/src/chainstate/coordinator/mod.rs
Comment (jcnelson) [reply stackslib/src/chainstate/coordinator/mod.rs:315]: Great, thanks!

File: stackslib/src/chainstate/nakamoto/coordinator/mod.rs
Comment (jcnelson) [reply stackslib/src/chainstate/nakamoto/coordinator/mod.rs:71]: Opened an issue: https://github.com/stacks-network/stacks-core/issues/4290

File: stackslib/src/chainstate/stacks/boot/mod.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/mod.rs:188]: Per @jferrant, it turns out that this actually needs to be a public key -- DKG needs the public keys before it can begin.

File: stackslib/src/chainstate/stacks/boot/mod.rs
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/mod.rs:653]: This is actually moot now that we know from @jferrant that we need the keys themselves.

**Review by jcnelson — COMMENTED** [review]
My subsequent review is blocked on the need for the signer set to track public keys instead of principals.

**Comment by kantai** [comment]
Okay -- this PR has now been updated with the work from #4295: the reward set is written to `.signers` _and_ the reward set entries contain the signer keys.

The signer keys are stored in the reward set (_not the stacking-state map_) as `(buff 33)`s, and checked for length. The signer keys are sent over the event dispatcher interface, in the consensus order used for the bit vec.

These signer keys have a many-to-many relationship with PoX addresses. Each reward set entry points to one signing key, but a PoX address can appear multiple times in the reward set entries, and different entries can point to the same signing key.

The `make_signer_set` sums over the signing keys, and this PR includes unit tests for its behavior.

File: testnet/stacks-node/src/event_dispatcher.rs
Comment (jferrant) [inline testnet/stacks-node/src/event_dispatcher.rs:75]: I see you can subscribe to new stacker signer sets (and not saying in this PR as this should get merged and this is orthogonal) but do ou think we should add an RPC endpoint for the current pox_set? We will need to retrieve them from the node at signer spin up so we can determine the signers and their vote shares.

File: testnet/stacks-node/src/config.rs
Comment (jferrant) [inline testnet/stacks-node/src/config.rs:2412]: I see we have the means to subscribe to stacker sets, but this doesn't help a signer that spins up after stackers have registered. Should I create a ticket to add this as an rpc endpoint as we will need to configure the signers with it?

**Comment by jferrant** [comment]
Aside from the failed build due to a missing bitvec in the one file, LGTM

**Review by jferrant — APPROVED** [review]

File: testnet/stacks-node/src/config.rs
Comment (kantai) [reply testnet/stacks-node/src/config.rs:2412]: Yes, please open a ticket -- it should be simple enough to add in a follow-up PR.

File: testnet/stacks-node/src/event_dispatcher.rs
Comment (zone117x) [inline testnet/stacks-node/src/event_dispatcher.rs:109]: @kantai did you intend to include this in the event payload? It looks to be unused right now

Comment (kantai) [reply testnet/stacks-node/src/event_dispatcher.rs]: Ah -- no, at one point I was going to, but it would have required some cloning of sizable structs (the reward set). Removed in 99730e0c6.

Comment (zone117x) [reply testnet/stacks-node/src/event_dispatcher.rs]: That data looked really helpful though! Right now the stacker list data in the event is just this this right?
```rs
pub struct RewardSet {
    pub rewarded_addresses: Vec<PoxAddress>,
    pub start_cycle_state: PoxStartCycleInfo,
}
```

File: testnet/stacks-node/src/config.rs
Comment (jferrant) [reply testnet/stacks-node/src/config.rs:2412]: Done https://github.com/stacks-network/stacks-core/issues/4298

File: testnet/stacks-node/src/event_dispatcher.rs
Comment (kantai) [reply testnet/stacks-node/src/event_dispatcher.rs]: It'll also contain:

```rust
   // only generated for nakamoto reward sets
   pub signers: Option<Vec<NakamotoSignerEntry>>,
```

Which will contain signing keys, stacked amount for each signing key, and the number of signing slots.

**Review by jcnelson — APPROVED** [review]

**Pull request merged**
